crates/solver/src/settlement_submission/submitter/eden_api.rs

```diff
@@ -1,37 +1,66 @@
 //! https://docs.edennetwork.io/for-traders/getting-started
 
-use crate::settlement::{Revertable, Settlement};
-
-use super::{
-    super::submitter::{TransactionHandle, TransactionSubmitting},
-    common::PrivateNetwork,
-    AdditionalTip, CancelHandle, Strategy, SubmissionLoopStatus,
+use crate::{
+    settlement::{Revertable, Settlement},
+    settlement_submission::{
+        submitter::{
+            common::PrivateNetwork, AdditionalTip, CancelHandle, Strategy, SubmissionLoopStatus,
+            TransactionHandle, TransactionSubmitting,
+        },
+        GlobalTxPool,
+    },
 };
+
 use anyhow::{bail, Context, Result};
 use ethcontract::{
     transaction::{Transaction, TransactionBuilder},
-    H256,
+    H160, H256, U256,
 };
 use futures::{FutureExt, TryFutureExt};
 use reqwest::{Client, IntoUrl, Url};
 use serde::Deserialize;
 use shared::{transport::http::HttpTransport, Web3, Web3Transport};
 use web3::{helpers, types::Bytes};
 
+const ALREADY_KNOWN_TRANSACTION: &[&str] = &[
+    "already known",
+    "nonce too low",
+    "replacement transaction underpriced",
+];
+
 #[derive(Clone)]
 pub struct EdenApi {
     client: Client,
     url: Url,
     rpc: Web3,
+    global_tx_pool: GlobalTxPool,
 }
 
 #[derive(Debug, Clone, Deserialize)]
 struct EdenSuccess {
     result: H256,
 }
 
+fn biggest_public_nonce(global_tx_pool: &GlobalTxPool, address: H160) -> Option<U256> {
+    let mut biggest_nonce = None;
+
+    for sub_pool in global_tx_pool.pools.lock().unwrap().iter() {
+        if !matches!(sub_pool.strategy, Strategy::CustomNodes) {
+            // we are only interested in transactions from the public mempool
+            continue;
+        }
+
+        for (sender, nonce) in sub_pool.pools.keys() {
+            if sender == &address && nonce > &biggest_nonce.unwrap_or_default() {
+                biggest_nonce = Some(*nonce);
+            }
+        }
+    }
+    biggest_nonce
+}
```

Comment:
Alternative in functional style. Don't have to do it this way, just an idea.
```suggestion
fn biggest_public_nonce_(global_tx_pool: &GlobalTxPool, address: H160) -> Option<U256> {
    let pools = global_tx_pool.pools.lock().unwrap();
    pools
        .iter()
        .filter(|sub_pool| matches!(sub_pool.strategy, Strategy::CustomNodes))
        .flat_map(|sub_pool| sub_pool.pools.keys())
        .filter(|(sender, _)| *sender == address)
        .map(|(_, nonce)| *nonce)
        .max()
}
```